packages/flutter_tools/lib/src/build_system/targets/font_subset.dart

```diff
@@ -0,0 +1,250 @@
+        }
+        result[familyKey] = fonts.first['asset'] as String;
+      }
+    }
+    return result;
+  }
+
+  Future<Map<String, List<int>>> _findConstants(File dart, File constFinder, File appDill) async {
+    final List<String> cmd = <String>[
+      dart.path,
+      constFinder.path,
+      '--kernel-file', appDill.path,
+      '--class-library-uri', 'package:flutter/src/widgets/icon_data.dart',
+      '--class-name', 'IconData',
+    ];
+    globals.printTrace('Running command: ${cmd.join(' ')}');
+    final ProcessResult constFinderProcessResult = await globals.processManager.run(cmd);
+
+    if (constFinderProcessResult.exitCode != 0) {
+      throwToolExit('ConstFinder failure: ${constFinderProcessResult.stderr}');
+    }
+    final dynamic jsonDecode = json.decode(constFinderProcessResult.stdout as String);
+    if (jsonDecode is! Map<String, dynamic>) {
+      throwToolExit('Invalid ConstFinder output: expected a top level JavaScript object, got $jsonDecode.');
+    }
+    final Map<String, dynamic> constFinderMap = jsonDecode as Map<String, dynamic>;
+    final _ConstFinderResult constFinderResult = _ConstFinderResult(constFinderMap);
+    if (constFinderResult.hasNonConstantLocations) {
+      globals.printError('This application cannot tree shake icons fonts. It has non-constant instances of IconData at the following locations:', emphasis: true);
+      for (final Map<String, dynamic> location in constFinderResult.nonConstantLocations) {
+        globals.printError('- ${location['file']}:${location['line']}:${location['column']}', indent: 2, hangingIndent: 4);
+      }
+      throwToolExit('Avoid non-constant invocations of IconData or try to build again without --shake-icon-fonts.');
+    }
+    return _parseConstFinderResult(constFinderResult);
+  }
+
+  Map<String, List<int>> _parseConstFinderResult(_ConstFinderResult consts) {
+    final Map<String, List<int>> result = <String, List<int>>{};
+    for (final Map<String, dynamic> iconDataMap in consts.constantInstances) {
+      if (iconDataMap['fontPackage'] is! String || iconDataMap['fontFamily'] is! String || iconDataMap['codePoint'] is! int) {
+        throwToolExit('Invalid ConstFinder result. Expected "fontPackage" to be a String, "fontFamily" to be a String, and "codePoint" to be an int, got: $iconDataMap.');
+      }
+      final String package = iconDataMap['fontPackage'] as String;
+      final String family = iconDataMap['fontFamily'] as String;
+      final String key = package == null
+        ? family
+        : 'packages/$package/$family';
+      result[key] ??= <int>[];
+      result[key].add(iconDataMap['codePoint'] as int);
+    }
+    return result;
+  }
+
+}
+
+class _ConstFinderResult {
+  const _ConstFinderResult(this.result);
+
+  final Map<String, dynamic> result;
+  List<Map<String, dynamic>> get constantInstances => _getList(result['constantInstances'], 'Invalid ConstFinder output: Expected "constInstances" to be a list of objects.');
```

Comment:
Cache the results of json parsing to avoid recomputing every time the getter is touched.